packages/test-runner-intercept/README.md

```diff
@@ -0,0 +1,141 @@
+# Web Test Runner Intercept
+
+Web test runner plugin that allows the interception of modules during test runs. Often, while creating tests, there is the need to change the behavior of a dependency of the system under test. This is needed to cover all branches of the system under test, for performance reasons or sometimes because the dependency cannot operate correctly in the test environment. Dependencies that are on the global window object are easy to change with a mock or a stub. ES Modules, however, and specifically their exports cannot be altered. This plugin introduces a hook that enables the interception of modules and allows altering its exports.
+
+## Concept
+
+Next to a web test runner plugin that enables module interception, this package also exposes a function `interceptModule()` to be used in the test run. This function allows to define what module needs to be intercepted and returns an object with all the exports of the module as its properties. Changing these properties allows to rewire the intercepted module without changing the system under test.
+
+```js
+import { interceptModule } from '@web/test-runner-intercept';
+
+const externalLibrary = await interceptModule('external-library');
+
+// Return the original function that 'external-library' exposed in the `externalFunction` named export
+externalLibrary.externalFunction; // f externalFunction() { ... }
+
+// Rewire the 'external-library' module to return this anonymous function as the `externalFunction` export
+externalLibrary.externalFunction = () => console.log('hello world'); // () => console.log('hello world')
+
+const { externalFunction } = await import('external-library');
+externalFunction; // () => console.log('hello world')
+```
+
+### Caveats
+
+#### Order of imports
+
+In order to intercept a module, the module should not be referenced yet. (Once a module is loaded, the module loader also loads all its dependent modules) This means the module containing the system under test should be loaded _after_ the module interception. As interception is achieved using a function, this also means the system under test should _always_ be loaded using a dynamic import in order to be done after the interception.
```

Comment:
It's not entirely clear to me what is referenced here as being "the system under test", can you clarify this? Or rather, what does this mean for when you're testing a library that you want to mock in a component?

Say you have a component, that in the connectedCallback calls some library that you want to "intercept". Does that mean that in your test file, you have to lazy load your component file?

Reply:
So, in your example, the "system under test" is the component. (https://en.wikipedia.org/wiki/System_under_test) That system is loaded through a module. (eg. `my-component.js` or `system-under-test.js` in the readme) In your example the component uses a library (eg. the default export of `external-library`), this is loaded using an import, often a static import.

In order for the test file to intercept the call to the library, it needs to do so before the library is loaded. So, this indeed means that in your test file you need to import the component module file **after** the interception. As the interception is a async/await function call, you indeed need to use a dynamic import in your test file after that interception call in order to reference/use the components in your test.

```js
import { interceptModule } from '@web/test-runner-intercept';

// First, intercept
const externalLibrary = await interceptModule('external-library');

// Second, load the system under test.
const systemUnderTest = await import('../../src/system-under-test.js');

// Run tests
...
```